gwpy/timeseries/timeseries.py

```diff
@@ -1476,11 +1488,48 @@ def whiten(self, fftlength, overlap=0, method='scipy-welch',
         scipy.signal
 
         Notes
-        -----"""
+        -----
+        If `fir=True` and `asd` is not `None`, the `fftlength`, `overlap`,
+        and `method` arguments are not used, and `window` determines the
+        shape of window used to truncate the whitening filter.
+
+        For more on the FIR whitening filter, see arXiv:gr-qc/0509116
+        """
         # build whitener
         if asd is None:
             asd = self.asd(fftlength, overlap=overlap,
                            method=method, window=window, **kwargs)
+        # if requested, process as an FIR filter
+        if fir:
+            # build the full filter
+            filter_length = int(filter_duration * self.sample_rate.value)
+            asd = asd.interpolate(1./self.duration.value)
+            invasd = 1. / asd.value
+            # zero the edges and apply a hard highpass
+            invasd[0] = 0
+            invasd[asd.size - 1] = 0
+            if highpass:
+                ncorner = int(highpass / asd.df.value)
+                invasd[0:ncorner] = 0
+            # truncate and window in the time domain
+            tdw = npfft.irfft(invasd)
+            trunc_start = int(filter_length / 2)
+            trunc_stop = self.size - int(filter_length / 2)
+            window = signal.get_window(window, filter_length)
+            tdw[0:trunc_start] *= window[int(filter_length/2):filter_length]
+            tdw[trunc_stop:self.size] *= window[0:int(filter_length/2)]
+            tdw[trunc_start:trunc_stop] = 0
+            # apply truncated filter in the frequency domain
+            invasd = numpy.abs(npfft.rfft(tdw))  # enforce zero-phase
+            in_ = self.detrend(detrend)
+            out = numpy.zeros(self.size).view(type(self))
+            out.__metadata_finalize__(self)
+            out._unit = self.unit
+            out.value[:] += npfft.irfft(in_.fft().value * invasd)
+            out.value[:] *= self.duration.value / 2  # re-normalise
```

Comment:
please combine these into a single operation (unless there's a good reason not to)